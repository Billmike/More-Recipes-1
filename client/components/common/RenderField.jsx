import React from 'react';
import PropTypes from 'prop-types';

const propTypes = {
  input: PropTypes.shape().isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  meta: PropTypes.shape({
    touched: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
};

/**
 * Input field component
 *
 * @param {object} Field - The input field
 * @param {object} Field.input - The input field element
 * @param {string} Field.label - The input field label
 * @param {string} Field.type - The input field type
 * @param {object} Field.meta
 * @param {boolean} Field.meta.touched - The field input state
 * @param {string} Field.meta.error - Validation error message
 *
 * @returns {object} Input element
 */
const RenderField = ({
  input,
  label,
  type,
  meta: { touched, error }
}) => (
  <div>
    <label htmlFor={label}>{label}</label>
    <br />
    <div>
      <input
        {...input}
        type={type}
        placeholder={label}
      />
      {touched && (error && <span className="red-text">{error}</span>)}
    </div>
  </div>
);

RenderField.propTypes = propTypes;

export default RenderField;
