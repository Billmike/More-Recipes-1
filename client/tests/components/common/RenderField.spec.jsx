import React from 'react';

import RenderField from '../../../components/common/RenderField';

describe('Render field component', () => {
  it('should render component', () => {
    const props = {
      input: {},
      type: 'text',
      label: 'Username',
      meta: {
        touched: false,
        error: null,
      }
    };
    const wrapper = mount(<RenderField {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
  it('should show error when present', () => {
    const props = {
      input: {},
      type: 'text',
      label: 'Username',
      meta: {
        touched: true,
        error: 'New error message',
      }
    };
    const wrapper = mount(<RenderField {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper.find('span.red-text').exists()).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });
});
