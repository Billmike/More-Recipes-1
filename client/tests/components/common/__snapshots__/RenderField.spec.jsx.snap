// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render field component should render component 1`] = `
<RenderField
  input={Object {}}
  label="Username"
  meta={
    Object {
      "error": null,
      "touched": false,
    }
  }
  type="text"
>
  <div>
    <label
      htmlFor="Username"
    >
      Username
    </label>
    <br />
    <div>
      <input
        placeholder="Username"
        type="text"
      />
    </div>
  </div>
</RenderField>
`;

exports[`Render field component should show error when present 1`] = `
<RenderField
  input={Object {}}
  label="Username"
  meta={
    Object {
      "error": "New error message",
      "touched": true,
    }
  }
  type="text"
>
  <div>
    <label
      htmlFor="Username"
    >
      Username
    </label>
    <br />
    <div>
      <input
        placeholder="Username"
        type="text"
      />
      <span
        className="red-text"
      >
        New error message
      </span>
    </div>
  </div>
</RenderField>
`;
